### ISSUE TITLE
Closing bracket uncovered in try-catch statement when exception is re-thrown in catch

### ISSUE THREAD

**Issue by bgoerdt** [tonerdo/coverlet#465]
The closing bracket is showing up as uncovered when I re-throw an exception from a `catch` statement. Is this expected behavior? The closing bracket is technically unreachable code, and it prevents me from reaching 100% coverage on the method when coverlet says it is uncovered by a test.

```
try
{
...
}
catch (Exception)
{
    // handle exception (retry/log/etc.)

    throw;
} // ** uncovered line **
```

**Comment by MarcoRossignoli** [comment]
Can you produce a simple repro?
Coverlet results are based on generated pdb(where we don't have control), to understand if it's a bug we would need a repro.

**Comment by brpratt** [comment]
This only seems to be an issue when there is an `await` inside the catch block. Here is a repro:

https://github.com/brpratt/CoverletAwaitThrowInsideCatch

From the [generated code](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwKwG5nJQDMmM6AogB4CGAtgA4A2ApsgN7LpeYlQAcmAGwAeAJYA7AC4A+dAAVqAJwDOzAILKAnuIDGACgDKkxRIDm6ZcYCU6Ttw5JuT9Mc13n6Bx49QA7OglJADoFFWY9S0UrfEdvAF93Zx1qSR0AC0SnL28nKABOISCAEWZGak09VGjM7xqPSTTFAHsAdxichNiuTrigA==), it seems there is a branch while deciding how to re-throw the exception. In the linked example, maybe this is the offender?

```cs
Exception ex = <>s__1 as Exception;
if (ex == null)
{
    throw <>s__1;
}
ExceptionDispatchInfo.Capture(ex).Throw();
```

**Comment by MarcoRossignoli** [comment]
Thank's for repro...I'll let you know what I discover.

**Comment by MarcoRossignoli** [comment]
did some check...and seem that result is expected for both cases sync and async, I mean if exceptio throw code rewind and so we cannot reach end of a method.
The only "incorrect" thing that I see is the branches on emtpy catch and on throw that is autogenerated, I don't know if I we can remove those noise without risk to remove "non false positive" I'll check what we can do.
TL;DR; The issue with this edge cases is that the pattern of compilers could change in future and if we're to much "aggressive" on "not so clear pattern recognition" we risk to intruduce bug where a branch/line is reported wronlgy covered.

![1](https://user-images.githubusercontent.com/7894084/66811521-89739480-ef31-11e9-9241-70c0b7dc0ad3.PNG)
![2](https://user-images.githubusercontent.com/7894084/66811526-8bd5ee80-ef31-11e9-8730-006a20d1b181.PNG)

**Comment by MarcoRossignoli** [comment]
Compiler generated code is below, at the moment I don't know how to recognize and filter branches out without remove other possible real branches.
Move issue level from bug to enhancement, I mean branches are real there is a mismatch between c# and IL.
```cs
// Coverlet.Core.Samples.Tests.CatchThrow
// Token: 0x06000021 RID: 33 RVA: 0x00002FF4 File Offset: 0x000011F4
public async Task<int> ParseAsync(string str)
{
	int num = 0;
	object obj2;
	try
	{
		return int.Parse(str);
	}
	catch (object obj)
	{
		obj2 = obj;
		num = 1;
	}
	int num2 = num;
	if (num2 == 1)
	{
		await Task.Delay(0);
		Exception ex = obj2 as Exception;
		if (ex == null)
		{
			throw obj2;
		}
		ExceptionDispatchInfo.Capture(ex).Throw();
	}
	obj2 = null;
	int result;
	return result;
}
```

**Comment by NoahApplebaum** [comment]
I am experiencing the same problem. Doesn't seem too much like an edge case because it happens in all cases where there is a return statement before closing brackets and some opening brackets.

Why is line 20 not considered for code coverage at all but line 21 is? 
Obviously there are other coverage issues in this picture, but that discrepancy jumps out at me the most. 

![image](https://user-images.githubusercontent.com/22988208/69441790-533be880-0d19-11ea-94fa-44942bf6b042.png)

**Comment by MarcoRossignoli** [comment]
@NoahApplebaum some questions, version of coverlet?Debug or release?Can you specify your "test" or better provide full repro(tests) with method above.

**Comment by NoahApplebaum** [comment]
Updated my coverlet.msbuild and it works! Thanks @MarcoRossignoli
